com.unity.shadergraph/Editor/Generation/Processors/PropertyCollector.cs

```diff
@@ -14,26 +16,186 @@ public struct TextureInfo
-            properties.Add(chunk);
+            }
+
+            int propIndex = -1;
+
+            if (m_ReferenceNames.TryGetValue(prop.referenceName, out propIndex))
+            {
+                // existing referenceName
+                var existingProp = m_Properties[propIndex];
+                if (existingProp.property != prop)
+                {
+                    // duplicate reference name, but different property instances
+                    if (existingProp.property.GetType() != prop.GetType())
+                    {
+                        Debug.LogError("Two properties with the same reference name using different types");
+                    }
+                    else
+                    {
+                        if (!EquivalentHLSLProperties(existingProp.property, prop))
+                            Debug.LogError("Two properties with the same reference name produce different HLSL properties");
+                    }
+                }
+                // set target flag to record who has declared this property
+                existingProp.declarerFlags = existingProp.declarerFlags | m_DeclarerFlag;
+                m_Properties[propIndex] = existingProp;
+            }
+            else
+            {
+                // new referenceName, new property
+                propIndex = m_Properties.Count;
+                m_Properties.Add(new Property() { property = prop, declarerFlags = m_DeclarerFlag });
+                m_ReferenceNames.Add(prop.referenceName, propIndex);
+            }
         }
 
-        public void GetPropertiesDeclaration(ShaderStringBuilder builder, GenerationMode mode, ConcretePrecision defaultPrecision)
+        public void ForEachPropertyUsedByTarget(Target target, Action<AbstractShaderProperty> action)
         {
-            foreach (var prop in properties)
+            // filter to properties declared by a node, graph, or the current Target
+            UInt64 declarerFlagFilter = (UInt64)DeclarerFlags._NodeOrGraph;
+            int targetIndex = m_Targets.FindIndex(t => t == target);
+            if (targetIndex >= 0)
+                declarerFlagFilter = declarerFlagFilter | (((UInt64)DeclarerFlags._Target) << targetIndex);
+
+            foreach (var p in m_Properties)
             {
-                // set up switched properties to use the inherited precision
-                prop.SetupConcretePrecision(defaultPrecision);
+                if (p.declarerFlags == 0)
+                    Debug.LogError("No Declarer! " + p.property.referenceName);
+                else if ((p.declarerFlags & declarerFlagFilter) != 0)
+                {
+                    action(p.property);
+                }
             }
+        }
 
-            // build a list of all HLSL properties
-            var hlslProps = new List<HLSLProperty>();
-            properties.ForEach(p => p.ForeachHLSLProperty(h => hlslProps.Add(h)));
+        private List<HLSLProperty> BuildHLSLPropertyListForTarget(Target target)
```

Comment:
This is used to build the HLSL property list for the specified Target.  It caches, so if you request the same Target, it doesn't have to build the HLSL property list again.